### PR TITLE
Bug fix of accum_grad when grad-nan

### PR DESCRIPTION
**Pull request by kamo-naoyuki**
Now optimizer.zero_grad() is not invoked when the gradient is nan and accum_grad>1.
This is a fatal bug of #3014 and probably causes different result from previous version... Maybe #3170 also comes from here.

### CODE REVIEW

**Comment by shincling** [comment]
Hi Kamo, I got one more question about the behavior of the Grad_nan related problem.  

If the grad-nan happens and there is accum_grad>1 in config, 
then the skip updating behavior will **_skip the specific batch (where nan happans)_** or **_the whole set of batches in this round of accum_grad_**?

I checked the trainer.py, but I'm not fully clear about this.

**Comment by kamo-naoyuki** [comment]
Hi, now, we selected `the whole set of batches in this round of accum_grad` because it is more similar behaviour to the real batch size without `accum_grad`.

Perhaps `skip the specific batch (where nan happans` is better, but actually, I don't know how to  implement it.

```
# iter1
loss.backward()  # grad = grad_iter1
# iter2
loss.backward()  # grad = grad_iter1 + grad_iter2
# iter3
loss.backward()  # grad = grad_iter1 + grad_iter2 + grad_iter3
...
```

Every time `loss.backward()`, gradients are accumulated, but I don't know how to get each gradient of each iteration before accumulated using PyTorch API, except for the first iteration.

**Comment by shincling** [comment]
> Hi, now, we selected `the whole set of batches in this round of accum_grad` because it is more similar behaviour to the real batch size without `accum_grad`.
> 
> Perhaps `skip the specific batch (where nan happans` is better, but actually, I don't know how to implement it.
> 
> ```
> # iter1
> loss.backward()  # grad = grad_iter1
> # iter2
> loss.backward()  # grad = grad_iter1 + grad_iter2
> # iter3
> loss.backward()  # grad = grad_iter1 + grad_iter2 + grad_iter3
> ...
> ```
> 
> Every time `loss.backward()`, gradients are accumulated, but I don't know how to get each gradient of each iteration before accumulated using PyTorch API, except for the first iteration.

Thanks, Kamo. That's quite helpful.
